.github/workflows/turing.yaml

```diff
@@ -491,44 +479,45 @@ jobs:
             --header "X-Vault-Token: ${{ env.VAULT_TOKEN }}" \
             --request POST \
             --data @credentials.json \
-            http://vault.127.0.0.1.nip.io/v1/secret/k3d-${{ env.CLUSTER_NAME }}
+            http://vault.127.0.0.1.nip.io/v1/secret/dev
 
+          # Create a secret for Merlin with information on how to access Vault
           kubectl create secret generic vault-secret \
-            --from-literal=address=http://vault:8200 \
-            --from-literal=token=${{ env.VAULT_TOKEN }}
-
-      - name: Install Merlin
-        env:
-          CLUSTER_NAME: k3d-${{ env.CLUSTER_NAME }}
-        run: |
-          helm install merlin ./merlin/charts/merlin \
-            --set merlin.image.tag=${{ env.MERLIN_VERSION }} \
-            --set merlin.environmentConfigs[0].cluster=${{ env.CLUSTER_NAME }} \
-            --set merlin.imageBuilder.clusterName=${{ env.CLUSTER_NAME }} \
-            --values infra/e2e/merlin.helm-values.yaml \
-            --wait
-
-          kubectl apply -f infra/e2e/merlin.ingress.yaml
-
-      - name: Install mockserver
-        run: |
-          kubectl apply -f infra/e2e/turing.mockserver.yaml
+            --from-literal=vault-address=http://vault:8200 \
+            --from-literal=vault-token=${{ env.VAULT_TOKEN }}
 
       - name: Install Turing
         run: |
           helm install turing ./infra/charts/turing \
-            --set turing.image.registry=${{ env.LOCAL_REGISTRY }}/ \
+            --values infra/e2e/turing.values.yaml \
+            --values infra/e2e/${{ matrix.valuesFile }} \
+            --set turing.image.registry=${{ env.LOCAL_REGISTRY }} \
             --set turing.image.repository=${{ github.repository }}/turing-api \
             --set turing.image.tag=${{ env.TURING_API_VERSION }} \
-            --set turing.config.RouterDefaults.Image=${{ env.LOCAL_REGISTRY }}/${{ github.repository }}/turing-router:${{ env.TURING_ROUTER_VERSION }} \
-            --values infra/e2e/turing.helm-values.yaml \
-            --wait
+            --set turing.config.RouterDefaults.Image=${{ env.LOCAL_REGISTRY }}/${{ github.repository }}/turing-router:${{ env.TURING_ROUTER_VERSION }}
 
-          kubectl apply -f infra/e2e/turing.ingress.yaml
+      - name: Install mockserver
+        run: |
+          kubectl apply -f infra/e2e/turing.mockserver.yaml
+
+      - uses: jupyterhub/action-k8s-await-workloads@v1
+        id: wait-for-deployment
+        with:
+          workloads: >-
+            deployment/mockserver,
+            deployment/turing-mlp,
+            deployment/turing-merlin,
+            deployment/turing
+          timeout: 300 #seconds
+          max-restarts: -1
 
       - if: steps.wait-for-deployment.outcome == 'failure' && always()
```

Comment:
What's this `always()` here ?

Reply:
`always()` allows running this step even if the previous step has failed. There is an implicit `success()` condition on each step (so step is not being executed if any of the previous steps has failed) hence it should be explicitly overridden here. I think it can also be replaced with `failure()`, but it doesn't change anything as we are also looking for another specific condition to happen (failure of the `wait-for-deployment` step).